### PR TITLE
Removed kafka integration, replaced by webhooks

### LINKED ISSUE #14: Résoudre les problèmes de crash du consumer hydra lors de l'envoi de beaucoup de messages kakfa en simultanée

### PR DESCRIPTION
**Pull request by quaxsze**
Fix https://github.com/etalab/data.gouv.fr/issues/982
Fix https://github.com/etalab/data.gouv.fr/issues/984
Fix #14

### CODE REVIEW

**Review by maudetes — COMMENTED** [review]
A first batch of review.

But I think I had the same logic in mind :)

File: udata_hydra/utils/http.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import requests
+
+from udata_hydra import config
+
+
+def send(dataset_id, resource_id, document):
+    uri = f'{config.UDATA_URI}/api/1/{dataset_id}/resources/{resource_id}/extras/'
+    headers = {'content-type': 'application/json', 'X-API-KEY': config.UDATA_URI_API_KEY}
+    r = requests.put(uri, headers=headers, json=document)
```
Comment (abulte) [inline udata_hydra/utils/http.py:9]: use aiohttp, this tool is based on it and it will leverage async :-)

**Review by abulte — COMMENTED** [review]
I wonder if we need a Celery or RQ to handle http calls to udata. Using aiohttp instead of requests will make it "less blocking" but still.

We'll need one scheduler anyway when adding analysis. But maybe we can wait till then.

File: udata_hydra/crawl.py [hunk 2]
```diff
@@ -104,26 +102,14 @@ async def update_check_and_catalog(check_data: dict) -> None:
...
-                        key_id=str(last_check["resource_id"]),
-                        document={**check_data, "check_date": str(datetime.now())},
-                        meta=meta,
-                    )
+                    log.debug("Sending message to Udata...")
+                    document = dict()
+                    document['check:status'] = check_data['status'] if status_has_changed else last_check["status"]
+                    document['check:timeout'] = check_data['timeout']
+                    document['check:check_date'] = str(datetime.now())
+                    send(dataset_id=last_check["dataset_id"],
```
[the remark this answers is not in the document]
Comment (quaxsze) [reply udata_hydra/crawl.py:110]:
```
            last_checks = [
                {
                    "resource_id": row[0],
                    "dataset_id": row[1],
                    "priority": row[2],
                    "initialization": row[3],
                    "status": None,
                    "timeout": None,
                }
                for row in rows
            ]
```
Looks like there is no?

[the remark this answers is not in the document]
Comment (maudetes) [reply udata_hydra/crawl.py]: This variable is poorly named? But `rows` come from `catalog` and not `check` table.

File: udata_hydra/utils/http.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+from udata_hydra import config
+
+log = logging.getLogger("udata-hydra")
+
+
+async def send(dataset_id: str, resource_id: str, document: dict) -> None:
+    if not config.UDATA_URI or not config.UDATA_URI_API_KEY:
+        log.error(g"Missing Udata URI and API key to send http query")
+        return
+    uri = f"{config.UDATA_URI}{dataset_id}/resources/{resource_id}/extras/"
```
Comment (maudetes) [inline udata_hydra/utils/http.py:13]:
```suggestion
    uri = f"{config.UDATA_URI}/datasets/{dataset_id}/resources/{resource_id}/extras/"
```

File: udata_hydra/utils/http.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    headers = {"content-type": "application/json", "X-API-KEY": config.UDATA_URI_API_KEY}
+    async with aiohttp.ClientSession() as session:
+        async with session.put(uri, json=document, headers=headers) as resp:
+            body = await resp.json()
```
Comment (maudetes) [inline udata_hydra/utils/http.py:17]: It fails if the content can't be parsed as `json`. Maybe you should print raw `content` if status != 200

File: udata_hydra/utils/http.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+
+async def send(dataset_id: str, resource_id: str, document: dict) -> None:
+    if not config.UDATA_URI or not config.UDATA_URI_API_KEY:
+        log.error(g"Missing Udata URI and API key to send http query")
+        return
+    uri = f"{config.UDATA_URI}{dataset_id}/resources/{resource_id}/extras/"
+    headers = {"content-type": "application/json", "X-API-KEY": config.UDATA_URI_API_KEY}
+    async with aiohttp.ClientSession() as session:
+        async with session.put(uri, json=document, headers=headers) as resp:
```
Comment (maudetes) [inline udata_hydra/utils/http.py:16]: `document` could have some `None` values, ex:
```
{
    'analysis:error': None,
}
```
However, udata extras can't be None.
We should either :
* allow None values in udata extras
* prevent sending None values here
* filter None values out on extras endpoint

File: udata_hydra/datalake_service.py [hunk 6]
```diff
@@ -99,58 +96,38 @@ async def process_resource(url: str, dataset_id: str, resource_id: str, response
...
             except ValueError:
                 log.debug(
                     f"Resource {resource_id} in dataset {dataset_id} is not a CSV"
                 )
 
-        # Send a Kafka message for both CSV and non CSV resources
+        # Send Udata a message for both CSV and non CSV resources
         log.debug(
-            f"Sending kafka message for resource analysed {resource_id} in dataset {dataset_id}"
+            f"Sending a message to Udata for resource analysed {resource_id} in dataset {dataset_id}"
```
Comment (maudetes) [inline udata_hydra/datalake_service.py:112]:
```suggestion
            f"Sending a message to udata for resource analysed {resource_id} in dataset {dataset_id}"
```

File: udata_hydra/crawl.py [hunk 7]
```diff
@@ -104,26 +102,14 @@ async def update_check_and_catalog(check_data: dict) -> None:
...
-                    }
-                    produce(
-                        kafka_uri=config.KAFKA_URI,
-                        topic=get_topic("resource.checked"),
-                        service="udata-hydra",
-                        key_id=str(last_check["resource_id"]),
-                        document={**check_data, "check_date": str(datetime.now())},
-                        meta=meta,
-                    )
+                    log.debug("Sending message to Udata...")
```
Comment (maudetes) [inline udata_hydra/crawl.py:105]:
```suggestion
                    log.debug("Sending message to udata...")
```

File: udata_hydra/datalake_service.py [hunk 8]
```diff
@@ -99,58 +96,38 @@ async def process_resource(url: str, dataset_id: str, resource_id: str, response
...
-                    resource_id,
-                    {
-                        "data_location": storage_location,
-                        "mime_type": mime_type,
-                        "filesize": filesize,
-                        "delimiter": delimiter,
-                        "encoding": encoding,
-                    },
-                    meta={"dataset_id": dataset_id, "message_type": "resource.stored"},
+                    f"Sending message to Udata for resource stored {resource_id} in dataset {dataset_id}"
```
Comment (maudetes) [inline udata_hydra/datalake_service.py:99]:
```suggestion
                    f"Sending message to udata for resource stored {resource_id} in dataset {dataset_id}"
```

File: udata_hydra/datalake_service.py [hunk 9]
```diff
@@ -99,58 +96,38 @@ async def process_resource(url: str, dataset_id: str, resource_id: str, response
...
+                await send(dataset_id=dataset_id,
+                           resource_id=resource_id,
+                           document=document)
             except ValueError:
                 log.debug(
                     f"Resource {resource_id} in dataset {dataset_id} is not a CSV"
                 )
 
-        # Send a Kafka message for both CSV and non CSV resources
+        # Send Udata a message for both CSV and non CSV resources
```
Comment (maudetes) [inline udata_hydra/datalake_service.py:110]:
```suggestion
        # Send udata a message for both CSV and non CSV resources
```

File: udata_hydra/utils/http.py [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+import logging
+
+from udata_hydra import config
+
+log = logging.getLogger("udata-hydra")
+
+
+async def send(dataset_id: str, resource_id: str, document: dict) -> None:
+    if not config.UDATA_URI or not config.UDATA_URI_API_KEY:
+        log.error(g"Missing Udata URI and API key to send http query")
```
Comment (maudetes) [inline udata_hydra/utils/http.py:11]:
```suggestion
        log.error(g"Missing udata URI and API key to send http query")
```

File: udata_hydra/utils/http.py [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+        return
+    uri = f"{config.UDATA_URI}{dataset_id}/resources/{resource_id}/extras/"
+    headers = {"content-type": "application/json", "X-API-KEY": config.UDATA_URI_API_KEY}
+    async with aiohttp.ClientSession() as session:
+        async with session.put(uri, json=document, headers=headers) as resp:
+            body = await resp.json()
+            if not resp.status == 200:
+                log.error(f"Udata reponded with a {resp.status} and content {body['message']}")
```
Comment (maudetes) [inline udata_hydra/utils/http.py:19]:
```suggestion
                log.error(f"udata reponded with a {resp.status} and content {body['message']}")
```

**Review by maudetes — COMMENTED** [review]
Some comments, notably on the relation with udata

File: udata_hydra/utils/http.py (hunk 5 above)
Comment (maudetes) [reply udata_hydra/utils/http.py:20]: Done in https://github.com/etalab/udata-hydra/pull/27/commits/5ecaba7dbba7ad3cd7661a2e8a352aa8ef1d848b

**Review by maudetes — APPROVED** [review]
Oh yeah! let's merge! :rocket:

**Pull request merged**